### ISSUE TITLE
Memory overhead of storing intermediate results

### ISSUE THREAD

**Issue by bhuvneshchaturvedi2512** [KULeuven-COSIC/vFHE#1]
Could you please let me know what is the amount of memory that is required to store all the intermediate results (of the homomorphic computations in the example) that are then used to generate the proofs? It is hard to follow/compute from the codebase.

**Comment by jannikspiessens** [comment]
I can do a quick back-of-the-envelope calculation. In that example, there are 3 ciphertexts with each 3 coefficients and a lattice dimension of 2^12 over a 120 bit modulus. So that comes to about 553kB.

**Comment by bhuvneshchaturvedi2512** [comment]
Thank you for your prompt response. If my understanding is correct, does the 3 coefficients represent the three finite fields? Moreover, in the paper, the BGV modulus is described as 3 primes of 30 bits, which should make the modulus be 90 bits and not 120 bits. Finally, I do not get how there are only three ciphertexts? Is it due to the first being the output of layer one computation, the second being the output of the relinearization, and the third being the output of layer two computation? I mean I do not understand how the abovementioned result is arrived.

**Comment by jannikspiessens** [comment]
In the intermediate layer, there are 3 ciphertexts since k_1 = 3 which are indeed each represented using 90 bits (30bits from each of the 3 primes). All of these ciphertexts have not yet been relinearized so they each have 3 coefficients (one of them is decomposed but this does not change the total number of bits required). Each of the ciphertexts coefficients is represented as a vector of dimension 2^12. So with using 90 instead of 120 bits, this comes to 415kB.

**Comment by bhuvneshchaturvedi2512** [comment]
Thank you for your prompt response. I have a few final (for the time being) questions that maybe you can help me with. In Section 5 of your paper, it is mentioned that "To simplify parameter selection for this specific instantiation, the moduli Q(0) and Q(1) contain an extra prime." Exactly how does this extra prime help in the simplification of the parameter selection? Moreover, in the same section, it is mentioned that "the proof size is about 187kB." Can you please help me understand how this figure was arrived at or computed? Finally, and correct me if I am wrong, the example computes a linear layer as v = a*c_{i} + b, where a is a weight matrix, c is the encrypted feature, and b is the bias vector. Once done, the activation function is computed as c_{i + 1} = v*v (the square in equation 2?). I believe since this last operation involves a multiplication, is the proof generated for this particular operation and not the one in the linear layer (since it only involves additions)? Also, do k0, k1, and k2 denote the dimensions of a and b at different layers? Again, thank you for your patience in answering my questions.

**Comment by jannikspiessens** [comment]
Let me answer your questions one by one:
- It ensures that the bounds for correctness in BGV hold. One could avoid this extra modulus by choosing primes larger than 30 bits.
- It was computed using libprocps.
- As can be seen in equation (2), k0, k1 and k2 denote the dimensions of the encrypted vectors for the input, intermediate and output layers respectively. All computations in a layer are proven. We start maintenance (relin/modswitch) after each square operations since this is a ciphertext-ciphertext multiplication while the linear operations do not involve ciphertext-ciphertext multiplications.

Feel free to ask further questions via email.

**Comment by bhuvneshchaturvedi2512** [comment]
Dear Jannik,

Thank you for your prompt response.

I just have one more question. Say I need the server to perform a single
homomorphic computation that consists of the ciphertext-ciphertext
multiplication between two RLWE ciphertexts (the size of each ciphertext
can be computed as 2*logQ*N). The server first performs the computation to
obtain a ciphertext with 3 polynomials (whose size is now 3*logQ*N) and
generates the associated proof. Once done, it performs the relinearization
operation to bring back the ciphertext to 2 polynomials (whose size is
again 2*logQ*N) and generates the associated proof. The question is, first,
what will be the structure of the associated proof for each computation (in
terms of polynomials), and second, what is the information that will be
sent back to the client to aid in verification. I believe the server needs
to send back the input ciphertexts and the final relinearized ciphertext
(all of which have 2 polynomials each), the intermediate ciphertext (which
has 3 polynomials), along with the proof of computations (both
multiplication and relinearization). Please correct me if I am wrong.

Awaiting your response.

Thanks & Regards
Bhuvnesh Chaturvedi
